### ISSUE TITLE
Errors on Make

### ISSUE THREAD

**Issue by k0z1can** [plashchynski/viewssld#10]
Hi Dmitry & Community,

Having some challenges compiling viewssld for use on Ubuntu server 14.04 , pasted below are the results from ./configure and make.

Any help would be greatly appreciated.

 root@ssl-decrypt:/etc/viewssld/viewssld-0.7.0# ./configure
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking target system type... x86_64-unknown-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/wait.h that is POSIX.1 compatible... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking arpa/inet.h usability... yes
checking arpa/inet.h presence... yes
checking for arpa/inet.h... yes
checking netinet/in.h usability... yes
checking netinet/in.h presence... yes
checking for netinet/in.h... yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking sys/ioctl.h usability... yes
checking sys/ioctl.h presence... yes
checking for sys/ioctl.h... yes
checking sys/socket.h usability... yes
checking sys/socket.h presence... yes
checking for sys/socket.h... yes
checking syslog.h usability... yes
checking syslog.h presence... yes
checking for syslog.h... yes
checking for unistd.h... (cached) yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/Makefile
config.status: creating doc/Makefile
config.status: creating distribute.sh
config.status: executing depfiles commands
config.status: executing default commands
checking pcap.h usability... yes
checking pcap.h presence... yes
checking for pcap.h... yes
checking openssl/ssl.h usability... yes
checking openssl/ssl.h presence... yes
checking for openssl/ssl.h... yes
checking dssl/sslcap.h usability... yes
checking dssl/sslcap.h presence... yes
checking for dssl/sslcap.h... yes
checking libnet.h usability... yes
checking libnet.h presence... yes
checking for libnet.h... yes

root@ssl-decrypt:/etc/viewssld/viewssld-0.7.0# make
Making all in src
make[1]: Entering directory `/etc/viewssld/viewssld-0.7.0/src'
gcc  -g -O2  -lpcap -lssl -lnet -ldssl -o viewssld viewssld.o utils.o
viewssld.o: In function`error_callback_proc':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:776: undefined reference to `SessionToString'
viewssld.o: In function`init_fake_session_state':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:408: undefined reference to `libnet_init'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:424: undefined reference to`libnet_seed_prand'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:437: undefined reference to `libnet_get_prand'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:438: undefined reference to`libnet_get_prand'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:431: undefined reference to `libnet_destroy'
viewssld.o: In function`sendEmptyTCPSegment':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:535: undefined reference to `libnet_build_tcp'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:559: undefined reference to`libnet_build_ipv4'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:583: undefined reference to `libnet_build_ethernet'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:595: undefined reference to`libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:535: undefined reference to `libnet_build_tcp'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:559: undefined reference to`libnet_build_ipv4'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:577: undefined reference to `libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:601: undefined reference to`libnet_write'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:610: undefined reference to `libnet_clear_packet'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:604: undefined reference to`libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:597: undefined reference to `libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:579: undefined reference to`libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:553: undefined reference to `libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:555: undefined reference to`libnet_geterror'
viewssld.o:/etc/viewssld/viewssld-0.7.0/src/viewssld.c:606: more undefined references to `libnet_geterror' follow
viewssld.o: In function`session_event_handler':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:447: undefined reference to `SessionToString'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:499: undefined reference to`pcap_stats'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:511: undefined reference to `libnet_destroy'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:465: undefined reference to`SessionSetCallback'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:467: undefined reference to `SessionSetMissingPacketCallback'
viewssld.o: In function`data_callback_proc':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:679: undefined reference to `libnet_build_tcp'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:713: undefined reference to`libnet_build_ipv4'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:737: undefined reference to `libnet_build_ethernet'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:749: undefined reference to`libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:765: undefined reference to `libnet_clear_packet'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:679: undefined reference to`libnet_build_tcp'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:713: undefined reference to `libnet_build_ipv4'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:707: undefined reference to`libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:765: undefined reference to `libnet_clear_packet'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:755: undefined reference to`libnet_write'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:758: undefined reference to `libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:731: undefined reference to`libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:709: undefined reference to `libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:751: undefined reference to`libnet_geterror'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:733: undefined reference to `libnet_geterror'
viewssld.o:/etc/viewssld/viewssld-0.7.0/src/viewssld.c:760: more undefined references to`libnet_geterror' follow
viewssld.o: In function `main':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:237: undefined reference to`SSL_library_init'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:238: undefined reference to `OpenSSL_add_all_ciphers'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:239: undefined reference to`OpenSSL_add_all_digests'
viewssld.o: In function `proceed':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:316: undefined reference to`pcap_open_live'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:328: undefined reference to `pcap_compile'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:333: undefined reference to`pcap_setfilter'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:342: undefined reference to `CapEnvCreate'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:344: undefined reference to`CapEnvSetSSL_ServerInfo'
viewssld.o: In function `main':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:243: undefined reference to`EVP_cleanup'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:244: undefined reference to `CRYPTO_cleanup_all_ex_data'
viewssld.o: In function`proceed':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:371: undefined reference to `CapEnvDestroy'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:377: undefined reference to`pcap_close'
viewssld.o: In function `main':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:243: undefined reference to`EVP_cleanup'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:244: undefined reference to `CRYPTO_cleanup_all_ex_data'
viewssld.o: In function`proceed':
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:357: undefined reference to `CapEnvSetSessionCallback'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:359: undefined reference to`CapEnvCapture'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:335: undefined reference to `pcap_geterr'
/etc/viewssld/viewssld-0.7.0/src/viewssld.c:330: undefined reference to`pcap_geterr'
collect2: error: ld returned 1 exit status
make[1]: **\* [viewssld] Error 1
make[1]: Leaving directory `/etc/viewssld/viewssld-0.7.0/src'
make: **\* [all-recursive] Error 1

Thank you,
Andrew

**Comment by crjang** [comment]
I met the same problem. undefined reference problem.
I already installed libdssl_2.1.1 and there are no any error messages when configuring...

Please let me know how can i solve it... Thanks.

**Comment by alecrwaters** [comment]
Hi,

 

Before carrying on working on getting it compiled, it’s worth thinking about this:

 

The issue that viewssld faces in today’s world is that more and more SSL/TLS servers are using some form of perfect forward secrecy (“PFS”, usually a flavour of Diffie-Hellman) for key agreement, rather than using the server’s private key to encrypt the pre-master secret. There’s lots of maths ‘n’ stuff here: http://vincent.bernat.im/en/blog/2011-ssl-perfect-forward-secrecy.html

 

The bottom line is that viewssld’s days are numbered unless you cripple your server’s SSL/TLS configuration to not use PFS (thereby scoring poorly on the SSL Labs test <https://www.ssllabs.com/ssltest/> !), so it (viewssld) doesn’t have much of a future outside of some very specific cases. I guess that in a PFS world the best way to do this would be to terminate SSL/TLS on a box in front of your servers and have your NSM tap on the unencrypted link between the two?

 

alec
